### PR TITLE
modularisation

### PR DESCRIPTION
**Pull request by sfranky**

### CODE REVIEW

File: pbs.py [hunk 1]
```diff
@@ -155,7 +155,7 @@ def make_qstat(orig_file, out_file, write_method):
     Some qstat files are structured a bit differently (the ones containing 'prior')
     Job id                    Name             User            Time Use S Queue
     or
-    job-ID  prior   name       user         node_dict submit/start at     queue                          slots ja-task-ID
+    job-ID  prior   name       user         ??????? submit/start at     queue                          slots ja-task-ID
```
Comment (fgeorgatos) [inline pbs.py]: what's the above? is it a mockup text? if so, good as-is

Comment (sfranky) [inline pbs.py]: yeap, that's how the input file looks like

File: pbs.py [hunk 2]
```diff
@@ -256,10 +256,10 @@ def read_pbsnodes_yaml(fn, write_method):
     return pbs_nodes
 
 
-def map_pbsnodes_to_wn_dicts(node_dict, pbs_nodes):
-    for (pbs_node, (idx, cur_node_nr)) in zip(pbs_nodes, enumerate(node_dict['wn_list'])):
-        node_dict['wn_dict'][cur_node_nr] = pbs_node
-        node_dict['wn_dict_remapped'][idx] = pbs_node
+def map_pbsnodes_to_wn_dicts(cluster_dict, pbs_nodes):
+    for (pbs_node, (idx, cur_node_nr)) in zip(pbs_nodes, enumerate(cluster_dict['wn_list'])):
```
Comment (fgeorgatos) [inline pbs.py]: what about renaming it to clsuter_dictionary ? code is written once but read many times over ... consider it

File: qtop.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
 from operator import itemgetter
 from optparse import OptionParser
 import datetime
-from collections import Counter, OrderedDict
+from collections import OrderedDict
```
Comment (fgeorgatos) [inline qtop.py]: the above is forward compatible, too? (ie. 3.x etc)

File: qtop.py [hunk 4]
```diff
@@ -47,20 +51,23 @@
...
+def colorize(text, pattern='Nothing', color_func=None, bg_colour=None):
     """
     prints text coloured according to a unix account pattern color.
+    If color is given, pattern is not needed.
     """
+    # bg_colour = code_of_color['BlueBG']
+    bg_colour = '' if not bg_colour else bg_colour
     try:
-        color_code = code_of_color[color] if color else code_of_color[color_of_account[pattern]]
+        ansi_color = code_of_color[color_func] if color_func else code_of_color[color_of_account[pattern]]
```
Comment (fgeorgatos) [inline qtop.py]: color_func: expand name maybe?

File: qtop.py (hunk 3 above)
Comment (sfranky) [inline qtop.py]: it is, but other stuff isn't !!

File: qtop.py [hunk 5]
```diff
@@ -47,20 +51,23 @@
...
     except KeyError:
         return text
     else:
-        return "\033[" + color_code + "m" + text + "\033[1;m" \
+        return "\033[" + '{}{}'.format(ansi_color, bg_colour) + "m" + text + "\033[0;m" \
             if ((not options.NOCOLOR) and pattern != 'account_not_coloured' and text != ' ') else text
 
 
-def decide_remapping(node_dict):
+def decide_remapping(cluster_dict, _all_letters, _all_str_digits_with_empties):
```
Comment (fgeorgatos) [inline qtop.py]: in python prepending a name with _ and __ have a special meaning; is this compliant usage?

File: qtop.py (hunk 4 above)
Comment (sfranky) [inline qtop.py]: for now it's a dummy with no purpose. This will be made in modularisation part two

File: qtop.py [hunk 6]
```diff
@@ -75,30 +82,37 @@ def decide_remapping(node_dict):
...
+    cluster_dict['all_digits'] = [int(digit) for digit in cluster_dict['all_str_digits']]
+
     if options.BLINDREMAP or \
-                    len(node_dict['node_subclusters']) > 1 or \
-                    min(node_dict['wn_list']) >= 9000 or \
-                    node_dict['offline_down_nodes'] >= node_dict['total_wn'] * config['percentage'] or \
-                    len(node_dict['_all_str_digits_with_empties']) != len(node_dict['all_str_digits']) or \
-                    len(node_dict['all_digits']) != len(node_dict['all_str_digits']):
+                    len(cluster_dict['node_subclusters']) > 1 or \
+                    min(cluster_dict['wn_list']) >= 9000 or \
```
Comment (fgeorgatos) [inline qtop.py]: hardwired value?

File: qtop.py (hunk 5 above)
Comment (sfranky) [inline qtop.py]: not really. Will remove in the future. Put it there mainly to make pycharm stop nagging about shadowing variables

File: qtop.py [hunk 7]
```diff
@@ -75,30 +82,37 @@ def decide_remapping(node_dict):
...
-    node_dict = dict()
+    cluster_dict = dict()
     for key in ['working_cores', 'total_cores', 'max_np', 'highest_wn', 'offline_down_nodes']:
-        node_dict[key] = 0
-    node_dict['node_subclusters'] = set()
-    node_dict['wn_dict'] = {}
-    node_dict['wn_dict_remapped'] = {}  # { remapnr: [state, np, (core0, job1), (core1, job1), ....]}
+        cluster_dict[key] = 0
+    cluster_dict['node_subclusters'] = set()
+    cluster_dict['wn_dict'] = {}
```
Comment (fgeorgatos) [inline qtop.py]: wn_dict: expand name?

File: qtop.py [hunk 8]
```diff
@@ -314,9 +331,9 @@ def fill_cpucore_columns(state_np_corejob, cpu_core_dict, id_of_username, max_np
     np = state_np_corejob['np']
     corejobs = state_np_corejob.get('core_job_map', '')
 
-    if state == '?':
-        for cpu_line in cpu_core_dict:
-            cpu_core_dict[cpu_line] += '_'
+    if state == '?':  # for non-existent machines
+        for core_line in core_user_map:
+            core_user_map[core_line] += ['#']  # was '_'. @Fotis Maybe closer to reality? (No machine, no cores!)
```
Comment (fgeorgatos) [inline qtop.py]: drop comment; can't recall the why of this; we'll be coming back to it, so make it a config tunable?

File: qtop.py [hunk 9]
```diff
@@ -339,31 +357,34 @@ def fill_cpucore_columns(state_np_corejob, cpu_core_dict, id_of_username, max_np
         these positions are filled with '#'s.
         '''
         for core in own_np_empty_range:
-            cpu_core_dict['Cpu' + str(core) + 'line'] += '_'
+            core_user_map['Core' + str(core) + 'line'] += ['_']
         for core in non_existent_cores:
-            cpu_core_dict['Cpu' + str(core) + 'line'] += '#'
-    return cpu_core_dict
+            core_user_map['Core' + str(core) + 'line'] += ['#']
```
Comment (fgeorgatos) [inline qtop.py]: good, but probably make them tunable characters (in the future, we may wish to replace with weird Unicode symbols!)

File: qtop.py (hunk 6 above)
Comment (sfranky) [inline qtop.py]: done

File: qtop.py (hunk 7 above)
Comment (sfranky) [inline qtop.py]: done

File: qtop.py (hunk 8 above)
Comment (sfranky) [inline qtop.py]: done

File: qtop.py [hunk 10]
```diff
@@ -426,38 +446,41 @@ def find_matrices_width(wn_number, wn_list, term_columns, DEADWEIGHT=11):
...
-        end_label = iter(end_labels[size])
-        for line_nr in wn_vert_labels:
-            highlight = highlight_alternately(*config['alt_label_highlight_colours'])
-            color_wn_id_list = list(line_with_separators(wn_vert_labels[line_nr][start:stop], SEPARATOR, options.WN_COLON))
-            color_wn_id_list = [colorize(elem, _, highlight.next()) for elem in color_wn_id_list]
-            line = ''.join(color_wn_id_list)
-            print line + end_label.next()
+        end_label = iter(end_labels[str(node_str_width)])
+        display_wnid_lines(wn_vert_labels, start, stop, end_label,
+                           color_func=highlight_alternately, args=(ALT_LABEL_HIGHLIGHT_COLOURS))
```
Comment (fgeorgatos) [inline qtop.py]: 4-space indent should be fine, no?

File: qtop.py [hunk 11]
```diff
@@ -468,16 +491,23 @@ def highlight_alternately(colour_a, colour_b):
...
-                                 DEADWEIGHT=11):
+def display_remaining_matrices(extra_matrices_nr,
+                               cluster_dict,
+                               core_user_map,
+                               print_end,
+                               pattern_of_id,
+                               wn_vert_labels,
+                               term_columns,
+                               wn_occup,
+                               DEADWEIGHT=11):
```
Comment (fgeorgatos) [inline qtop.py]: yeah, 4-space indent everywhere plz!

File: qtop.py [hunk 12]
```diff
@@ -468,16 +491,23 @@ def highlight_alternately(colour_a, colour_b):
...
-                                 term_columns,
-                                 DEADWEIGHT=11):
+def display_remaining_matrices(extra_matrices_nr,
+                               cluster_dict,
+                               core_user_map,
+                               print_end,
+                               pattern_of_id,
+                               wn_vert_labels,
+                               term_columns,
+                               wn_occup,
```
Comment (fgeorgatos) [inline qtop.py]: wn_occupancy?

File: qtop.py [hunk 13]
```diff
@@ -468,16 +491,23 @@ def highlight_alternately(colour_a, colour_b):
...
-                                 cpu_core_dict,
-                                 _print_end,
-                                 pattern_of_id,
-                                 wn_vert_labels,
-                                 term_columns,
-                                 DEADWEIGHT=11):
+def display_remaining_matrices(extra_matrices_nr,
+                               cluster_dict,
+                               core_user_map,
+                               print_end,
```
Comment (fgeorgatos) [inline qtop.py]: unclear naming

File: qtop.py [hunk 14]
```diff
@@ -486,26 +516,56 @@ def calculate_remaining_matrices(node_state,
...
+        'wn id lines': (print_wnid_lines, (print_start, print_end, cluster_dict['highest_wn'], wn_vert_labels),
+                        {'inner_attrs': None}),
+        'core user map': (print_core_lines, (core_user_map, print_start, print_end, pattern_of_id), {'attrs': None}),
+        # 'node state': (print_single_attr_line, (print_start, print_end), {'attr_line': wn_occup['node state']}),
+        # 'temperature': (print_single_attr_line, (print_start, print_end), {'attr_line': wn_occup['temperature']}),
+    }
+    for yaml_key, part_name in get_yaml_key_part('wn_occupancy'):
+        occupancy_parts.update(
+            {part_name: (print_single_attr_line, (print_start, print_end), {'attr_line': wn_occup[part_name]})}
+        )
```
Comment (fgeorgatos) [inline qtop.py]: for readability: collapse three lines into two, by assigning fi. `newdictvar` 'h kapoio kalytero onoma)

File: qtop.py [hunk 15]
```diff
@@ -535,76 +595,111 @@ def create_user_accounts_pool_mappings(account_jobs_table):
...
-                SEPARATOR,
-                options.WN_COLON
-            )
-        )
-        color_cpu_core_list = [colorize(elem, pattern_of_id[elem]) for elem in color_cpu_core_list if elem in pattern_of_id]
-        line = ''.join(color_cpu_core_list)
-        yield line + colorize('=Core' + str(ind), 'account_not_coloured')
+    for ind, k in enumerate(core_user_map):
+        cpu_core_line = core_user_map['Core' + str(ind) + 'line'][print_start:print_end]
+        if options.REM_EMPTY_CORELINES and '#' * (print_end - print_start) == cpu_core_line:
```
Comment (fgeorgatos) [inline qtop.py]: where possible, avoid hardwired values

File: qtop.py [hunk 16]
```diff
@@ -535,76 +595,111 @@ def create_user_accounts_pool_mappings(account_jobs_table):
...
 
-    for _node in node_dict['wn_dict']:
-        state_np_corejob = node_dict['wn_dict'][_node]
-        _cpu_core_dict = fill_cpucore_columns(state_np_corejob, _cpu_core_dict, id_of_username, max_np_range, user_of_job_id)
+    return _core_user_map
 
-    return _cpu_core_dict
 
+def calc_attr():
+    pass
```
Comment (fgeorgatos) [inline qtop.py]: whut's this?

Comment (fgeorgatos) [inline qtop.py]: override by user func?

File: qtop.py [hunk 17]
```diff
@@ -535,76 +595,111 @@ def create_user_accounts_pool_mappings(account_jobs_table):
...
-def calculate_wn_occupancy(node_dict, user_names, job_states, job_ids):
+
+def get_yaml_key_part(major_key):
+    """
+    only return the list items of the yaml major_key if a yaml key subkey exists
+    (this signals a user-inserted value)
+    """
+    # e.g. major_key = 'wn_occupancy'
+    for part in config[major_key]:
+        part_name = [i for i in part][0]
```
Comment (fgeorgatos) [inline qtop.py]: [0] implies something like a key? just confirm this is appropriate type of data structure

File: qtop.py (hunk 10 above)
Comment (sfranky) [inline qtop.py]: done

File: qtop.py (hunk 12 above)
Comment (sfranky) [inline qtop.py]: done

File: qtop.py [hunk 18]
```diff
@@ -670,13 +765,14 @@ def load_yaml_config(path):
...
     """
-    Calculation of split screen size
+    Calculates where to break the matrix into more matrices, because of the window size.
     """
     try:
         _, term_columns = os.popen('stty size', 'r').read().split()  # does not work in pycharm
     except ValueError:  # probably Pycharm's fault
         # _, term_columns = [52, 211]
         _, term_columns = [53, 176]
+        # _, term_columns = [53, 126]  # for smaller screens, namely 1366 x 768
```
Comment (fgeorgatos) [inline qtop.py]: oops, magic valuesm why not tunable?

File: qtop.py (hunk 13 above)
Comment (sfranky) [inline qtop.py]: renamed to print_char_stop

File: qtop.py [hunk 19]
```diff
@@ -688,7 +784,9 @@ def calculate_split_screen_size():
     QTOPPATH = os.path.expanduser('~/PycharmProjects/qtop')  # qtoppath: ~/qtop/qtop
```
Comment (fgeorgatos) [inline qtop.py]: best option for configs is likely: `~/.local/qtop`

File: qtopconf.yaml [hunk 20]
```diff
@@ -12,41 +9,80 @@ user_colour_mappings:  # order should be from more generic-->more specific
...
+     # customised end labels
+     end_labels:
+       '1': ['={__WNID__}']
+       '2': ['={_Worker_}', '={__Node__}']
+       '3': ['={_Worker_}', '={__Node__}', '={___ID___}']
+       '4': ['={________}', '={_Worker_}', '={__Node__}', '={___ID___}']
+       '5': ['={__FOR___}', '={__SALE__}', '={_Worker_}', '={__Node__}', '={___ID___}']
+       '6': ['={_SPACE__}', '={__FOR___}', '={__SALE__}', '={_Worker_}', '={__Node__}', '={___ID___}']
+       '7': ['={__PLACE_}', '={__YOUR__}', '={_ADVERT_}', '={__HERE__}', '={_Worker_}', '={__Node__}', '={___ID___}']
+ - temperature:
```
Comment (fgeorgatos) [inline qtopconf.yaml]: eithet temperature does not belong to wn_occupancy or, one of the labels is misleading (likely the first one)

File: qtopconf.yaml [hunk 21]
```diff
@@ -12,41 +9,80 @@ user_colour_mappings:  # order should be from more generic-->more specific
...
+     label: Mojo-ness
+     max_len: 1
+ - core user map:
+     options1: {}
+     options2: {}
+
+
+
+
+
```
Comment (fgeorgatos) [inline qtopconf.yaml]: remove too many extra spaces; LGTM!!!

File: qtopconf.yaml [hunk 22]
```diff
@@ -12,41 +9,80 @@ user_colour_mappings:  # order should be from more generic-->more specific
...
+#Gray_D Gray_L
+#Green Green_L
 #normal
 #Pink
 #Purple
-#Red
-#Red_L
+#Red Red_L
 #White
 #Yellow
```
Comment (fgeorgatos) [inline qtopconf.yaml]: an extra CR;LF here

File: qtop.py (hunk 17 above)
Comment (sfranky) [inline qtop.py]: get the key of the only key:value pair in the dictionary.
necessary evil given the ordered structure I chose for the yaml conf file

File: qtop.py (hunk 3 above)
Comment (fgeorgatos) [inline qtop.py]: ok, super ;-)

**Pull request merged**